### PR TITLE
pd: remove get leader addr

### PR DESCRIPTION
**Pull request by huachaohuang**
ref #942 #956

### CODE REVIEW

File: src/pd/client.rs [hunk 1]
```diff
@@ -72,17 +76,81 @@ fn parse_urls(addr: &str) -> Result<Vec<String>> {
...
+    }
+
+    if hosts.len() > 0 {
+        Ok(hosts)
+    } else {
+        Err(box_err!("invalid pd members object"))
+    }
+}
+
+fn get_pd_members(endpoints: &str) -> Result<Vec<String>> {
```
Comment (siddontang) [inline src/pd/client.rs]: do we need hyper to get PD members?
I think only the PD endpoints passed by argument is enough.

Comment (huachaohuang) [inline src/pd/client.rs]: What if we change PD members?

Comment (siddontang) [inline src/pd/client.rs]: This is a rare event, e,g, if we pass 3 PD endpoints, we can't remove this 3 PD and add another 3 PD, this is crazy in production. 

Btw, we can get all members in RPC, so maybe we can first choose an endpoint to connect, and then get all members. But I think we can support this later, now using argument passing is ok. 

I still think whether we should get leader or not first, or use proxy directly? how do you think?

Comment (huachaohuang) [inline src/pd/client.rs]: We can get rid of hyper if we don't need to get leader, and the implementation will be simpler.

File: src/server/node.rs [hunk 2]
```diff
@@ -189,6 +191,19 @@ impl<C> Node<C>
...
 
+    fn check_cluster_bootstrapped(&self) -> Result<bool> {
+        for _ in 0..MAX_CHECK_CLUSTER_BOOTSTRAPPED_RETRY_COUNT {
+            match self.pd_client.is_cluster_bootstrapped() {
+                Ok(b) => return Ok(b),
+                Err(e) => {
+                    warn!("check cluster bootstrapped failed: {:?}", e);
+                }
+            }
+            thread::sleep(Duration::from_millis(100));
```
Comment (siddontang) [inline src/server/node.rs]: should longer, 1s or more is ok.
define a const too.

**Comment by siddontang** [comment]
Rest LGTM

File: src/pd/client.rs [hunk 3]
```diff
@@ -82,7 +71,8 @@ fn rpc_connect(addr: &str) -> Result<TcpStream> {
         };
 
         // Send a HTTP header to tell PD to hijack this connection for RPC.
-        let header = b"GET /pd/rpc HTTP/1.0\r\n\r\n";
+        let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
+        let header = header_str.as_bytes();
         try!(stream.set_write_timeout(Some(Duration::from_secs(SOCKET_WRITE_TIMEOUT))));
         match stream.write(header) {
```
Comment (BusyJay) [inline src/pd/client.rs]: Use `write_all` instead.

Comment (huachaohuang) [inline src/pd/client.rs]: Will that return error on timeout?

Comment (BusyJay) [inline src/pd/client.rs]: Yes.

File: src/pd/client.rs [hunk 4]
```diff
@@ -82,39 +71,26 @@ fn rpc_connect(addr: &str) -> Result<TcpStream> {
...
-            Ok(n) => {
-                return Err(box_err!("write header failed: header len {} written len {}",
-                                    header.len(),
-                                    n));
-            }
-            Err(e) => {
-                return Err(box_err!("write header failed: {:?}", e));
-            }
-        }
+        try!(stream.write_all(header));
```
Comment (BusyJay) [inline src/pd/client.rs]: Do we need to try all the remaining hosts before returning error?

**Comment by BusyJay** [comment]
Now we have a local pd-server now, can the mock-pd be removed/replaced?

**Comment by siddontang** [comment]
seem not now, we inject some special logic in mock-pd for test. Maybe not easy to be replaced.

File: src/pd/client.rs [hunk 5]
```diff
@@ -58,63 +59,40 @@ fn send_msg(stream: &mut TcpStream, msg_id: u64, message: &Request) -> Result<(u
...
-                                    n));
-            }
-            Err(e) => {
-                return Err(box_err!("write header failed: {:?}", e));
-            }
+        let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
+        let header = header_str.as_bytes();
+        match stream.write_all(header) {
+            Ok(_) => return Ok(stream),
+            Err(_) => continue,
```
Comment (BusyJay) [inline src/pd/client.rs]: A warning may be better.

**Comment by BusyJay** [comment]
rest LGTM

**Pull request merged**